전체적인 구조 변경

GangWoon opened the pull request:
struct 기반으로 코드를 작성하기 위해서 노력했습니다. 중복되는 부분을 protocol로 묶으려고 노력하다보니 속성들을 private로 사용하기 위해서 protocol자체를 private로 주면서 작성하고 다른 빈 프로토콜을 상속받아서 다형성을 사용하려고 했지만 강제로 다운캐스팅을 하면서 원래 가지고 있던 값들을 날려버려서 포기하고 구조체 속성들을 private가 아니지만 직접 접근하지 않는 방식으로 도전해보겠습니다.
struct를 사용하려고 했는데 변경할때 더 이상 해결방법을 찾지 못해서 클래스로 전향했습니다.

```
gamers.forEach { gamer in
gamer의 속성을 바꾸는 메소드
} 
```
 게이머가 구조체인 경우 에러가 발생하는걸 처리할 수 없었습니다.

GangWoon committed:
프로토콜 파일 생성

GangWoon committed:
카드 구조체 생성

기본으로 필요한 속성을 만들었습니다.

GangWoon committed:
게이머 클래스 추가

핸즈에 카드를 추가할 수 있도록 만들었습니다.

GangWoon committed:
딜러 클래스 생성

기본 구조만 만들었습니다.

GangWoon committed:
게임 클래스 생성

스타일을 받아서 개수만큼 카드를 나눠 가질수 있도록 만들었습니다.
프로토콜을 채택해서 열거형의 원시값을 사용하지 않고 description이라는 값을 사용하도록 했습니다.

GangWoon committed:
손패 구조체 수정

테스트를 하기 위해서 클로저를 추가했습니다.

GangWoon committed:
게이머 클래스 수정

테스트 코드를위해서 수정했습니다. 클로저로 받은 배열을 출력합니다.

GangWoon committed:
게임 클래스 수정

테스트 코드 작성을 위해서 추가했습니다.

GangWoon committed:
게임 클래스 수정

테스트를 위해서 수정

GangWoon committed:
테스트 코드 완성~

godrm commented on CardGameApp/CardGameApp/Model/Game.swift:
```diff
@@ -0,0 +1,58 @@
+    var players: Int {
+        return gamers.count
+    }
+    private var gamers: [Gamer] = []
```
players 는 보기에 count 같지 않고 gamers 배열 같은 느낌이네요. 좀 더 명확하게 표현해도 좋겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/Game.swift:
```diff
@@ -0,0 +1,58 @@
+        case .five: return 5
+        case .seven: return 7
+        }
+    }
```
이 부분은 아직 억지로 변환해서 넘기는 것 같습니다. 
enum 타입을 switch-case로 다른 값으로 넘기는 것은 조심해서 접근하세요. 
타입을 만들면 그 타입을 최대한 활용하려고 노력해보세요

godrm replied:
enum 타입을 없애버린 건 좀 아쉽네요. Int 타입을 쓰는 것보다는 enum 타입이 구체적인 역할을 하기 때문에 더 좋은 선택입니다.
제 피드백의 요지는 enum 타입을 쓰려면 Int 타입인 rawValue에 접근하지 말고 그 자체를 활용하려고 노력하라는 겁니다. 참고하세요.

GangWoon replied:
더 생각해보겠습니다~

godrm commented on CardGameApp/CardGameApp/Model/Gamer.swift:
```diff
@@ -0,0 +1,32 @@
+    func printCard() {
+        cards.printCard { result in
+            print(result)
+        }
```
데이터나 로직을 포함하는 객체가 print- 계열 함수를 갖고 있는 것은 좋지 않습니다. 
항상 값을 넘겨서 다른 곳에서 출력하게 될 겁니다. 
이 함수는 앱을 만들때는 전혀 쓸모가 없는 게 됩니다. 심지어 테스트도 불가능한 함수입니다.

GangWoon replied:
테스트 코드를 위한게 아닌 제 눈으로 확인하기 편하게 만들었던 코드였습니다. 말씀하신대로 테스트를 하는 용도엔 적합하지 않는 코드가 맞습니다. 다음 쳅터로 넘어가면 지우도록 하겠습니다~

godrm commented:
객체끼리 협력 관계는 적당해 보입니다. 피드백을 확인해보세요.

GangWoon committed:
Player 구조체 생성

게이머를 배열로 같고 있을 필요가 없어서 player라는 구조체를 만들었습니다.

GangWoon committed:
프로토콜 삭제

접근법을 프로토콜이 아닌 타입 자체를 활용해서 사용하기 위해서 프로토콜을 삭제 시켰습니다.

GangWoon committed:
OCP 원칙 적용

확장은 개방적이고, 수정은 폐쇄적으로 만들고 싶었습니다.
열거형의 원시값이나 스위치문을 사용해서 값을 접근하는게 아닌 다른 방법을 생각했습니다.

GangWoon committed:
Player 구조체 수정

player를 구조체로 변경하고, game 구조체에서 속성을 사용할 수 있게 만들었습니다.

GangWoon committed:
Player 구조체 수정

외부에서 게이머를 추가할 수 있게 만들었습니다.

GangWoon committed:
Player 구조체 수정

이름이 이상해서 수정했습니다. 잘못 작성한 코드도 고쳤습니다.

GangWoon committed:
Game 구조체 수정

주석 처리했던 메소드를 활성화 시켰습니다.  style의 원시값이 삭제시키고 limitCount라는 속성을 사용해서 카드 배분하는걸 처리했습니다.

GangWoon committed:
컨벤션 맞추기

GangWoon committed:
테스트 작성을 위한 작업

player가 몇명인지 확인할 수 있도록 count 속성 추가

GangWoon committed:
테스트 작업을 위한 작업

game 구조체에 gamersCount라는 속성 추가

GangWoon committed:
테스트 코드 작성

GangWoon committed:
Player 이름 수정

players가 적절할 거 같아서 변경했습니다.

GangWoon commented:
허허... ocp에 대해서 다시 정리하고 고치겠습니다 지금 코드에선 Style이란 enum이 필요없어서 아침까지 더 고치겠습니다~

GangWoon committed:
Style 열거형 삭제

style 열겨형을 사용할 이유를 못 느껴서 삭제했습니다.

GangWoon commented:
protocol을 사용해서 ocp원칙을 사용할 가 생각해봤었습니다. 그런데 지금 구조에서 style 열거형을 제거하고 사용하는거와 protocol을 사용해서 나누는거랑 다를게 없어 보였습니다.

godrm commented:
그럼 PR을 더 열어놓고 수정하겠다는 의미인가요?

GangWoon commented:
아침에 코드 반영했습니다~

godrm commented:
다음 단계로 진행해보세요.
